flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/operators/AbstractAsyncStateStreamOperatorV2.java

```diff
@@ -260,6 +262,15 @@ public void processWatermark(Watermark mark) throws Exception {
         asyncExecutionController.processNonRecord(() -> super.processWatermark(mark));
     }
 
+    protected void emitWatermarkDirectly(Watermark mark) throws Exception {
+        if (timeServiceManager != null) {
+            CompletableFuture<Void> future = timeServiceManager.advanceWatermark(mark);
+            future.thenAccept(v -> output.emitWatermark(mark));
```

Comment:
I think we should also consider `EpochManager.ParallelMode` here. If serial, should wait & drain here.